### DISCUSSION TITLE
Is client multithreaded?

### DISCUSSION THREAD

**Discussion by gaurav2001trader** [miguelgrinberg/python-socketio#1008] [General]
```py
condition = true
def onMessage():
    global conditon
    if condition: 
       # condition = False ---> This works
       # do something 
       condition = False
```
By my logic onMessage should only run once. However onMessage gets executed a couple of times before stopping when triggered multiple times in a very short period of time. If I set `condition = False`  just after I checked the boolean it works fine.  Is this because of multithreading?  I am using `socketio.Client`

**Comment by miguelgrinberg** [comment]
Yes, the client is multithreaded.

**Reply by gaurav2001trader** [reply]
So how can I solve my problem. How can I prevent race condition in the threads

**Reply by miguelgrinberg** [reply]
I can't give you an answer specific to the Socket.IO client because you haven't shown a single line of code that is specific to this client, all you shared is a regular Python function. From a standard Python point of view you can use the usual synchronization primitives, maybe a `Lock` will work for you.

**Reply by gaurav2001trader** [reply]
https://github.com/symphonyfintech/xts-pythonclient-api-sdk/blob/main/MarketDataSocketClient.py
https://github.com/symphonyfintech/xts-pythonclient-api-sdk/blob/main/MarketdataSocketExample.py

I was writing the previous logic inside `on_message1502_json_full`. Now `on_message1502_json_full` is triggered continuously  and race condition is created. What I want is the code inside `on_message1502_json_full` is only executed by one thread at a time

**Reply by miguelgrinberg** [reply]
My reply above stands. Use a `Lock`.

**Reply by gaurav2001trader** [reply]
Ok I understand how to use lock with threads but how do I access the thread while using socketIO library. One way I can think is spawn threads inside the `on_message1502_json_full` and target them to some other function where I can lock the threads.

**Reply by miguelgrinberg** [reply]
You don't need to access the thread. Following your example:

```python
condition = true
myLock = Lock()

def onMessage():
    global condition
    with myLock:
        if condition: 
           # do something 
           condition = False
```

**Reply by gaurav2001trader** [reply]
Ok thankyou for your time
